### ISSUE TITLE
Remoting: Not getting reference to primary replica.

### ISSUE THREAD

**Issue by zeodamian** [Azure/service-fabric-issues#413]
We've had this problem before, and I've written about it [here (issue #192)](https://github.com/Azure/service-fabric-issues/issues/192). Lately it has been showing again.

It is very hard to see why this happens, we're just calling the ServiceProxy as normal, and all of a sudden  for some partitions a reference to a secondary replica is returned instead of to a primary.
We've even had to implement a restart function, that calls
`await client.ServiceManager.RestartReplicaAsync(_ctx.NodeName, _ctx.PartitionId, _ctx.ReplicaId);`
every time this happens.

Is it normal to need this kind of check for every method in a stateful service's interface?

```
async Task<Result> IFooService.DoWriteOperation()
        {
            if (this.Partition.ReadStatus != PartitionAccessStatus.Granted)
            {
                await TryRestartReplica();
                return Result.BadRequest($"Not primary replica ({this.Partition.ReadStatus}). Restarting FooService replica..");
            }
            return await _state.DoWriteOperation();
        }
```

The services are naturally fetched from ServiceProxy class with TargetReplicaSelector.PrimaryReplica.

Is there some way to make sure that I always get a reference to a Primary replica?
Shall I do some sort of wrapper around the call to ServiceProxy, and do some initial checking against the fabric state? Any suggestions appreciated.

**Comment by zeodamian** [comment]
I thought for a while that it was the primary being recycled by the fabric, but I just noticed that the fabric had - at least once - kept the primary running, and returned the reference to a secondary. Then after I had restarted that secondary replica (by asking for a reference to the _primary_ and asking it to restart), I again got the reference to the replica that had all the time been primary.

I could know it had all the time been primary (with readstatus granted) because it was running a process that would have been aborted if it was restarted, and it was still returning the increasing progress.

I am quite puzzled by this.

**Comment by zeodamian** [comment]
Maybe I should describe the broader picture:

We have a cluster where we let services download state from a blob once. The state is saved to reliable dictionaries in the stateful services, that we call seed stores, which then populate actor and service states.
We then run some operations in the system, and when we are done, we reset state in all reliable collections and actors, except the reliable dictionaries called seed stores. Once state has been downloaded from blobs, it is accessible from these seed stores instead, for faster iteration. So we can reset and reinitialize the cluster with the same base state, upon which we then enact scenarios which produces various states.

So, there are 10 partitions of one service for example, where the largest number of entities being loaded into reliable dictionaries from these seed stores, are round 140 k.
There are around 120 k actor instances being activated with state, and then deleted upon the reset.

The progress of this is followed by querying the services (the partitions) for progress, which is saved in it's state. Both the reset of state, and the reloading of state is monitored.

There are 2 service types managing different states. One of them is capable of resuming the state initialisation even when it is aborted in the midst of it. The other is not.
So when suddenly the fabric is not returning a reference to the primary for that service, and we can no longer monitor the progress, it becomes problematic.

We have to reset the state again, and reinitialize. Maybe several times within a short timespan. 

It seems like it leaves the services less capable of handling the following load, since eventually we will see these secondary replication limit reached errors (which did not show when services where freshly deployed).

Is there maybe some advice we can get here how to not run into these problems?

**Comment by zeodamian** [comment]
It seems the Fabric is doing a poor job of keeping the services available and healthy.

After redeploying and leaving it over night, FabricObjectClosedExceptions is thrown when trying to write to it's state.
For some reason, the replica acting as primary is not in a state to be written to, and fabric has not detected this and switched it out for a healthy replica. Ok, so we do this manually by restarting the primary replica. And then, we are back in this never ending story of having to restart the primaries because we every other time get a reference to a secondary.

The strain on the cluster, and the number of problems that arise from constantly getting a secondary instead of a primary, and only for short times getting the primary, are quite high.

It has been tried to restart all nodes and delete data and redeploy services, also to reinstall the cluster.

But this inability to retreive a reference to a primary replica is something that persists.

**Comment by appi101** [comment]
Does the cluster have any health warnings or errors when you hit this situation for this service? Does it happen during cluster or application upgrades?

**Comment by vturecek** [comment]
@zeodamian It's possible to lose read/write status at any time, but it should be rare and very brief under normal circumstances. When this occurs, is your cluster showing any health warnings or errors? Also can you link to other issues you've opened here for reference?

**Comment by zeodamian** [comment]
No errors or warnings, no upgrades going on.

Exact same code works locally, but in the cloud cluster getting references to the primary replica is like trying to catch a ghost. Most common to appear after putting some amount of strain on the cluster.

This is generally hard to reproduce as it seems to be coming after some heavy usage.

**Comment by zeodamian** [comment]
Some update:

Even locally this problem appears, but it has been to a lesser extent.

After resetting local cluster from tray icon, it has actually behaved perfectly.

**Cloud cluster, how is it reset? I would like to try this on it too.**
Because right now, every 5 minutes or so, there's an actor that tries to call a stateful service, and cannot get a reference to primary. We have retrial on it (even though it seems that if it happens even once, then we should restart immediately), and after a couple of times we let the stateful service replica restart itself.
It shouldn't be like this.

**Comment by VipulM-MSFT** [comment]
@zeodamian - can we get an update on this issue.

The ServiceProxy does not automatically refreshes the resolution unless there is  an error. What transport are you using for the service?

**Comment by VipulM-MSFT** [comment]
There was a scenario in which a primary replica when changes role to secondary may still continue to receive the requests if the service never throws an exception. This issue was fixed in 5.7.221 runtime SDK 2.7.221 and above. Please let us know if you still see an issue.

**Comment by VipulM-MSFT** [comment]
Closing this issue as per the above comment. If you still see the issue in the new runtime and SDK, please re-open this issue or create another one.
